### PR TITLE
JDK16 fixes and workarounds that should make the EA job pass

### LINKED ISSUE #16195: CI: replace JDK 15 with 16
## Description
JDK 16 reached GA a few weeks ago and the natural cycle would be to replace the jobs in Quarkus CI that are currently running on JDK 15 with JDK 16.
Our EA test job/workflow [1] is still running on 16 (and not 17-ea) because there are still some problems, especially with Kotlin and Gradle.
The idea is to get the EA job to the point where we can make the switch (both for CI JDK 15 -> 16 _and_ EA job 16 -> 17-ea).

[1] https://github.com/quarkusio/quarkus/actions/workflows/jdk-early-access-build.yml (see also #15867)

FTR:
- https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Gradle.207
- https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Quarkus.20tested.20on.20Java.2016
- https://groups.google.com/g/quarkus-dev/c/AlyXp2A2W5Y

/cc @gsmet @gastaldi 

### Currently known TODOs (potentially incomplete)

- [x] fix or work around [KT-43704 `Illegal reflective access by com.intellij.util.ReflectionUtil to method java.util.ResourceBundle.setParent(java.util.ResourceBundle) `](https://youtrack.jetbrains.com/issue/KT-43704)
  - [ ] update to Kotlin 1.5 (see also: #16806)
  - [ ] **or** (try to) skip all Kotlin bits in the entire repo for now
  - [x] **or** set `--add-opens java.base/java.util=ALL-UNNAMED` via `MAVEN_OPTS`
    - [x] stop `BootstrapMavenOptions` from passing `--add-opens` etc. to Maven `CliManager` (which chokes on that), see also #16862
    _note: complete **HACK** currently (in my fork), must be done properly_
- [x] work around [Gradle 6 not being compatible with JDK16+](https://github.com/gradle/gradle/issues/13481)
  - [x] don't execute `devtools/gradle` and `its/gradle`
  - [x] exclude anything related to Gradle from `devtools/cli` tests
- [x] mitigate XStream problem in its/main: https://github.com/quarkusio/quarkus/issues/16195#issuecomment-830872415
- [x] mitigate IAE in elytron-security-ldap: https://github.com/quarkusio/quarkus/issues/16195#issuecomment-830874787
- [x] add workaround for Surefire `java.io.tmpdir` problem: https://github.com/quarkusio/quarkus/issues/16195#issuecomment-830860621
- [x] fix inoker failures in its/container-image
- [x] check/fix remaining "Integration Tests - Devtools" failures
- [x] check/fix "Integration Tests - Kotlin" failures
- [x] check/fix TCKS failures
- [ ] ...?

For now checked TODOs can be found in my fork: https://github.com/quarkusio/quarkus/compare/main...famod:jdk16-fixes

### PR DESCRIPTION
**Pull request by famod**
First step for #16195.
See also: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Quarkus.20tested.20on.20Java.2016/near/237676424

I tried to divide everything in commits than can be addressed individually later, e.g. when fault-tolerance is fixed that one commit could be reverted.

I think the most controversial commit is the first one with `--add-opens`.
After a chat with @aloubyansky I decided to at least make sure it isn't propagated to tests via `MAVEN_OPTS` (which might spawn their own `mvn` subprocesses).
I suggest to keep `excludedEnvironmentVariables` _permanently_ because I think we generally don't want that propagation (until further notice).
I'm currently testing a `--illegal-access=warn` alternative to that `--add-opens` but I'm not convinced it's that much better.

For that `tmpdir` commit see https://github.com/quarkusio/quarkus/issues/16195#issuecomment-830860621 and https://github.com/quarkusio/quarkus/issues/16195#issuecomment-831523787. It's a bit of a mess currently but I didn't see a better solution until something in `jboss-parent` and/or Surefire changes.

### CODE REVIEW

**Comment by famod** [comment]
Just pinging a few people to take notice that there are things to fix for JDK16:
- @evanchooly for Kotlin; note that we already have #16806
- @glefloch for Gradle; there was a discussion on Zulip about toolchains
- @geoand for xstream
- @Ladicek for fault-tolarance; we already have https://github.com/quarkusio/quarkus/issues/16195#issuecomment-832623719
- ? for rest-client TCKS...maybe @michalszynkiewicz ?
- not sure who feels responsible for `elytron-security-ldap`?

I'm planning on creating issues for each "area" but I really don't mind if you create those issues yourself. 😉

File: integration-tests/elytron-security-ldap/pom.xml [hunk 1]
```diff
@@ -99,5 +99,15 @@
...
             </properties>
         </profile>
+        <profile>
+            <id>jdk16-workarounds</id>
+            <activation>
+                <jdk>[16,)</jdk>
+            </activation>
+            <properties>
+                <!-- avoids IllegalAccessException "java.naming does not export com.sun.jndi.ldap" -->
+                <surefire.argLine.additional>--add-exports java.naming/com.sun.jndi.ldap=ALL-UNNAMED</surefire.argLine.additional>
```
Comment (famod) [inline integration-tests/elytron-security-ldap/pom.xml:109]: A shoot...I missed that one.

**Comment by Tibor17** [comment]
@famod 
I introduced `excludedEnvironmentVariables` exactly because we in Apace Maven also had problems with `MAVEN_OPTS` in our Jenkins builds.

`<excludedEnvironmentVariables>MAVEN_OPTS</excludedEnvironmentVariables>`

**Comment by famod** [comment]
@Tibor17 thanks, good to know!

**Comment by geoand** [comment]
> * @geoand for xstream

I remember you mentioned that a new xstream version that fixes this problem is in the works, right?

**Comment by famod** [comment]
@geoand I hope things will be better with 1.5 but no clue when that will be released.
I'll give current `master` a try next week or so.

**Comment by aloubyansky** [comment]
No objections @famod Thanks!

**Review by gastaldi — APPROVED** [review]
LGTM, I'll let @gsmet in charge of pushing the big green merge button 😉

**Comment by famod** [comment]
EA job passed completely in my fork: https://github.com/famod/quarkus/actions/runs/819992102

**Review by gsmet — APPROVED** [review]
Nice work, thanks a lot for this!

**Pull request merged**